File: src/main/java/org/example/compiler/Function.java
package org.example.compiler;

import org.antlr.runtime.tree.CommonTree;
import org.antlr.runtime.tree.Tree;
import org.example.WhileLexer;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import static org.example.compiler.Compiler.compile;
import static org.example.compiler.Compiler.getChildren;

public class Function implements Element {


    public static LinkedList<String> parseOutput(Tree tree) {
        if (tree.getType() != WhileLexer.OUTPUTS) {
            throw new RuntimeException("Expected OUTPUTS node");
        }
        LinkedList<String> result = new LinkedList<>();
        ArrayList<Tree> outputs = getChildren(tree);
        for (Tree output : outputs) {
            result.add(output.getText());
        }
        return result;
    }

    public static void toCode(List<String> result, Tree tree) {

        String name = tree.getChild(0).getText();
        Tree output = tree.getChild(3);
        LinkedList<String> outputParsed = parseOutput(output);
        Tree body = tree.getChild(2);


        result.add("func begin %s".formatted(name));
        compile(body, result);
        for (String s : outputParsed) {
            result.add("return " + s);
        }
        result.add("Return");
        result.add("func end");
    }
}

File: src/main/java/org/example/compiler/Function.kt
package org.example.compiler

import org.antlr.runtime.tree.Tree
import org.example.WhileLexer

import java.util.*

object Function : Element {
    private fun parseOutput(tree: Tree): LinkedList<String> {
        if (tree.type != WhileLexer.OUTPUTS) {
            throw RuntimeException("Expected OUTPUTS node")
        }
        val result = LinkedList<String>()
        val outputs = Compiler.getChildren(tree)
        for (output in outputs) {
            result.add(output.text)
        }
        return result
    }

    private fun parseInput(tree: Tree): LinkedList<String> {
        if (tree.type != WhileLexer.INPUTS) {
            throw RuntimeException("Expected INPUTS node")
        }
        val result = LinkedList<String>()
        val inputs = Compiler.getChildren(tree)
        for (input in inputs) {
            result.add(input.text)
        }
        return result
    }

    @JvmStatic
    fun toCode(result: MutableList<String>, tree: Tree) {
        val name = tree.getChild(0).text
        val inputs = tree.getChild(1)
        val inputsParsed = parseInput(inputs)
        val output = tree.getChild(3)
        val outputParsed = parseOutput(output)
        val body = tree.getChild(2)
        result.add("func begin $name")
        for (input in inputsParsed) {
            result.add("get $input")
        }
        Compiler.compile(body, result)
        for (s in outputParsed) {
            result.add("return $s")
        }
        result.add("Return")
        result.add("func end")
    }
}
